### PR TITLE
Fix lora training for time-related parameters

### PR DESCRIPTION
**Pull request by thrfirs**
## Issue

The issue with the original code is that it only checks whether a module name contains `".time_"` when the `enable_time_finetune` option is enabled. However, in RWKV-v4neo, the only module that contains `".time_"` in its name is `time_shift` in `RWKV_TimeMix`, which is not trainable because it is a `nn.ZeroPad2d`.

As a result, time-related trainable `nn.Parameters` such as `time_first` in `RWKV_TimeMix` and `time_mix_k` in `RWKV_ChannelMix` will never be set to `requires_grad = True` since they are not modules.

## Fix

To address this issue, this pull request takes into account parameter names in `module.named_parameters()` when the `enable_time_finetune` option is enabled. This ensures that gradients are enabled for time-related parameters.

### CODE REVIEW

File: RWKV-v4neo/src/model.py [hunk 1]
```diff
@@ -62,32 +69,31 @@ def forward(ctx, B, T, C, w, u, k, v):
...
+            @staticmethod
+            def backward(ctx, gy):
+                B = ctx.B
+                T = ctx.T
+                C = ctx.C
+                gw = torch.empty((C), device=gy.device, memory_format=torch.contiguous_format, dtype=torch.bfloat16)
+                gu = torch.empty((C), device=gy.device, memory_format=torch.contiguous_format, dtype=torch.bfloat16)
+                gk = torch.empty((B, T, C), device=gy.device, memory_format=torch.contiguous_format, dtype=torch.bfloat16)
+                gv = torch.empty((B, T, C), device=gy.device, memory_format=torch.contiguous_format, dtype=torch.bfloat16)
+                return (None, None, None, gw, gu, gk, gv)
```
Comment (Blealtan) [inline RWKV-v4neo/src/model.py:94]: Why removed backward operator? Why are these gradients left empty?

File: RWKV-v4neo/src/model.py [hunk 2]
```diff
@@ -62,32 +69,31 @@ def forward(ctx, B, T, C, w, u, k, v):
...
 else:
-    wkv_cuda = load(name=f"wkv_{T_MAX}", sources=["cuda/wkv_op.cpp", "cuda/wkv_cuda.cu"], verbose=True, extra_cuda_cflags=["-res-usage", "--maxrregcount 60", "--use_fast_math", "-O3", "-Xptxas -O3", "--extra-device-vectorization", f"-DTmax={T_MAX}"])
+    wkv_cuda = load(name=f"wkv_{T_MAX}", sources=["cuda/wkv_op.cpp", "cuda/wkv_cuda.cu"], verbose=True, extra_cuda_cflags=["-res-usage", "--maxrregcount 40", "--use_fast_math", "-O3", "-Xptxas -O3", "--extra-device-vectorization", f"-DTmax={T_MAX}"])
```
Comment (Blealtan) [inline RWKV-v4neo/src/model.py:96]: Can you explain why the reg count is adjusted? I was leaving it unchanged from the Bo's original impl.

File: RWKV-v4neo/src/model.py [hunk 3]
```diff
@@ -151,21 +151,20 @@ def RUN_CUDA(B, T, C, w, u, k, v):
...
             "alpha"], LORA_CONFIG["dropout"]
-        self.lora_A = nn.Parameter(torch.empty(r, in_features))
-        self.lora_B = nn.Parameter(torch.empty(out_features, r))
+        self.lora_A = nn.Parameter(torch.empty(r, in_features, dtype=DTYPE))
+        self.lora_B = nn.Parameter(torch.empty(out_features, r, dtype=DTYPE))
         self.lora_dropout = nn.Dropout(dropout)
         self.scaling = alpha / r
 
-        nn.init.kaiming_uniform_(self.weight, a=math.sqrt(5))
+        nn.init.zeros_(self.weight)
```
Comment (Blealtan) [inline RWKV-v4neo/src/model.py:167]: The initialization assumes the base model is uninitialized so a Kaiming init is applied as is in a usual linear layer. It's not really necessary though.

File: RWKV-v4neo/cuda/wkv_cuda.cu [hunk 4]
```diff
@@ -2,8 +2,8 @@
 #include <assert.h>
 
 #define MIN_VALUE (-1e38)
+typedef float F;
```
Comment (Blealtan) [inline RWKV-v4neo/cuda/wkv_cuda.cu:5]: Why fix it to FP32? FP16/BF16 have been working fine.

File: RWKV-v4neo/src/model.py [hunk 5]
```diff
@@ -151,21 +151,20 @@ def RUN_CUDA(B, T, C, w, u, k, v):
 
 
 class LoraLinear(nn.Module):
-
     def __init__(self, in_features: int, out_features: int, bias: bool):
         super().__init__()
 
-        self.weight = nn.Parameter(torch.empty((out_features, in_features)))
+        self.weight = nn.Parameter(torch.empty((out_features, in_features), dtype=DTYPE), requires_grad=False)
```
Comment (Blealtan) [inline RWKV-v4neo/src/model.py:157]: If I remember it correctly, the `dtype` is globally specified in the Lightning wrapper. Should not need to give it here explicitly.

File: RWKV-v4neo/src/model.py [hunk 6]
```diff
@@ -535,21 +534,15 @@ def forward(self, idx):
...
 
         if args.tiny_att_dim > 0:
-            for block in self.blocks:
-                if args.grad_cp == 1:
-                    if args.lora:
-                        x = torch_checkpoint(block, x, x_emb, use_reentrant=False)
-                    else:
-                        x = deepspeed.checkpointing.checkpoint(block, x, x_emb)
+            for i, block in enumerate(self.blocks):
+                if i and args.grad_cp == 1:
```
Comment (Blealtan) [inline RWKV-v4neo/src/model.py:538]: Do you mean that `--grad_cp 1` works lora and deepspeed on your side? It didn't work for many others as well as myself, and I've looked into the implementation and confirmed it's the `use_reentrant` switch that matters. How did you fix that?

File: RWKV-v4neo/src/trainer.py [hunk 7]
```diff
@@ -18,6 +18,9 @@ class train_callback(pl.Callback):
     def __init__(self, args):
         super().__init__()
         self.args = args
+    
+    def on_train_start(self, trainer, pl_module):
```
Comment (Blealtan) [inline RWKV-v4neo/src/trainer.py:22]: These things remained unchanged compared to BlinkDL's original implementation. I don't expect to change those in this fork; you might want to submit another PR to the upstream repository.

File: RWKV-v4neo/train.py [hunk 8]
```diff
@@ -122,8 +122,7 @@
     import numpy as np
     import torch
     from torch.utils.data import DataLoader
-    if "deepspeed" in args.strategy:
-        import deepspeed
```
Comment (Blealtan) [inline RWKV-v4neo/train.py:126]: This modification is intentional so that deepspeed is loaded on-demand since it doesn't work on Windows.

File: RWKV-v4neo/train.py [hunk 9]
```diff
@@ -256,9 +258,10 @@
     os.environ["RWKV_JIT_ON"] = "1"
     if "deepspeed_stage_3" in args.strategy:
         os.environ["RWKV_JIT_ON"] = "0"
-    if args.lora and args.grad_cp == 1:
-        print('!!!!! LoRA Warning: Gradient Checkpointing requires JIT off, disabling it')
-        os.environ["RWKV_JIT_ON"] = "0"
+    
+    os.environ["RWKV_EMPTY_BACKWARD"] = "0"
+    if args.strategy == "deepspeed_stage_1":
```
Comment (Blealtan) [inline RWKV-v4neo/train.py:263]: Why is empty backward legal with deepspeed stage 1?

File: RWKV-v4neo/train.py [hunk 10]
```diff
@@ -304,15 +313,19 @@
...
                     for pname, param in module.named_parameters():
                         param.requires_grad = 'lora_' in pname
-                elif ((enable_time_finetune and '.time_' in name)
-                        or (enable_ln_finetune and '.ln' in name)):
-                    print(f'  LoRA additionally training {name}')
+                elif enable_ln_finetune and '.ln' in name:
+                    rank_zero_info(f'  LoRA additionally training module {name}')
                     for param in module.parameters():
                         param.requires_grad = True
+                elif enable_time_finetune and any(n.startswith("time") for n, _ in module.named_parameters()):
```
Comment (Blealtan) [inline RWKV-v4neo/train.py:324]: Finally it's the part you claim in the title...

**Review by Blealtan — CHANGES_REQUESTED** [review]
Thank you for your enormous effort in refining the implementation here and there. However, this PR bloats way more than you claim in the title, and many modifications in the PR might better go directly upstream. I've commented on the confusing points throughout your changes; please either
1. rename the PR to a broader name and convince me on every point, or
2. close the PR and reopen another one containing only the time-related modification.

Thanks again for your contribution!

**Comment by thrfirs** [comment]
Sincerely sorry for the confusion. Regrettably, I pushed a PR of my main branch by mistake, so all of my commits were exposed here while I had only intended to contribute a single commit.
As many of my modifications only work for my personal use, I will be closing this PR and creating a new one that contains only the time-related bug fix.

I am truly sorry for any inconvenience this has caused. And thank you for your kind and patient review of this PR!

**Comment by thrfirs** [comment]
Just for reference, I will explain my other modifications here. If you think these changes are necessary or interesting, please tell me and I will open separate PRs for them.

1. **`backward` and CUDA related modifications**: In LoRA training, only the `forward` for RWKV kernel is used, so I thought it would be better to optimize its speed instead of balancing its performance with `backward`. Actually, I removed the `backward` CUDA code completely so that the `forward` kernel could use all GPU register resources at compile time. Also I made it compatible with deepspeed stage 2 and 3 by setting empty parameters to match the shape requirements of the API. But surely, if we want to be compatible with upstream, it would be better to separate LoRA training settings into another set of CUDA codes, instead of directly modifying the existing ones. (By the way, I'm not sure why empty backward works for deepspeed stage 1 either; It just worked in my experiments)
2. **Zero/empty initilizations and explicit `dtype`s**: These changes were made to improve the model building procedure. Zero initialization makes it faster, and explicit `dtype`s make it less RAM consuming. Please note that this was done before pytorch_lightning trainer could specify the `dtype` of the model, which was too late to reduce the RAM cost for model building. Again, these should not be rewritten directly; the better way is to use if-else controls to enable them for LoRA only.
3. **LoRA+deepspeed+`grad_cp==1`**: Yes, somehow I made it work! The real issue with `deepspeed.checkpointing.checkpoint` for LoRA is that, just as its doc-string tells us, it works exactly like `torch.utils.checkpoint.checkpoint` with `use_reentrant=True`, which [requires](https://pytorch.org/docs/stable/checkpoint.html) "at least one of the inputs needs to have :code:`requires_grad=True` if grads are needed for model inputs". Unfortunately, in LoRA training, the direct input just before our LoRA linear layers is the frozen word embedding, which does not and shouldn't have grads. This is why your original code failed with both LoRA & deepspeed's grad checkpoint. By allowing the first LoRA linear layer to forward without being checkpointed, we can make all other layers' checkpoints work properly with native deepspeed with only a small VRAM cost increase. This makes the checkpointing mildly faster because it does not need to comply with `use_reentrant=False`.
4. **Loggings and other minor changes**: They are just for my personal use, and should not be included here.

I hope you will find these explanations useful! Let me know if you have any further questions.

**Comment by Blealtan** [comment]
1. Why in LoRA training only `wkv_forward` is used? I thought the gradients of `k` and `v` needed to be backpropagated to the previous KV `matmul`s, since we need the gradient to update the LoRA weights. After all, LoRA is just an approximation of full fine-tuning, and if we don't need the gradients in LoRA fine-tuning, we can also drop them during real full training, which is incorrect. But your experiments seem to be against these. Could you please explain that in more detail?
2. A cleaner way is to skip model weight initialization during `LoraLinear.__init__` and initialize `lora_A/B` on `set_state_dict`. As we are doing LoRA, we know it will be a base or LoRA model loaded before real computation. When it's loaded, we can see if `lora_A/B` are already in the `state_dict`: if not, we initialize them according to the type of base weight. I'll later work it out to see the VRAM gain; thanks for your notes.
3. That's a really awesome and clean solution. I may have missed that part of the modification, though. Please open a PR for this if you have the time. Thank you!
